nni/compression/pytorch/quantization_speedup/integrated_tensorrt.py

```diff
@@ -44,346 +38,320 @@ def valid_config(config=None):
             a_bits = config[name]['output_bits']
             assert a_bits in support_bits, "output bits should be 8, 16, 32"
 
-def handle_gemm(network, layer_idx, config):
-    """
-    This function handles special gemm operation due to layer numbers of gemm changed during pytorch->onnx model convertion.
+def print_layer_precisions(network):
+    print('The layer precisions and dynamic ranges are:')
+    for layer_idx in range(network.num_layers):
+        layer = network.get_layer(layer_idx)
+        out = layer.get_output(0)
+        print(layer.name, layer.precision, out.dynamic_range)
 
-    Parameters
-    ----------
-    network : tensorrt.INetworkDefinition
-        Represents a TensorRT Network from which the Builder can build an Engine
-    layer_idx : int
-        layer index of gemm
-    config : dict
-        Config recording bits number and name of layers
+def _handle_gemm(layer, config, out2layer, in2layer):
+    """
+    Gemm is special case. the following is the graph structure of Gemm in trt's graph
+    input                       ->| Gemm  ->| ElementWise
+    LayerType.Constant (weight) ->|
+    LayerType.Constant (bias) -> Shuffle  ->|
+    assume quantize input, output, and weight
     """
-    layer = network.get_layer(layer_idx)
-    pre_layer = network.get_layer(layer_idx-1)
-    next_layer = network.get_layer(layer_idx+1)
-    # if weight bits exists, set three layers' precision,
-    # input tensor range and the first two layers' output type
-    if 'weight_bits' in config[layer.name]:
-        assert 'tracked_min_input' in config[layer.name]
-        assert 'tracked_max_input' in config[layer.name]
-        w_bits = config[layer.name]['weight_bits']
-        tracked_min_input = config[layer.name]['tracked_min_input']
-        tracked_max_input = config[layer.name]['tracked_max_input']
-        # set three layers the same precision
+    w_bits = config['weight_bits']
+    layer.precision = Precision_Dict[w_bits]
+    # handle the input tensor
+    in_tensor = layer.get_input(0)
+    in_tensor.dynamic_range = (config['tracked_min_input'], config['tracked_max_input'])
+    # handle the output tensor
+    out_tensor = layer.get_output(0)
+    out_tensor.dynamic_range = (config['tracked_min_output'], config['tracked_max_output'])
+    # handle weight
+    w_in_tensor = layer.get_input(1)
+    weight_layer = out2layer[w_in_tensor.name]
+    assert weight_layer.type == trt.LayerType.CONSTANT
+    weight_layer.precision = Precision_Dict[w_bits]
+    weight_layer.set_output_type(0, Precision_Dict[w_bits])
+    w_out_tensor = weight_layer.get_output(0)
+    w_out_tensor.dynamic_range = (config['min_weight'], config['max_weight'])
+    print('special gemm: ', w_out_tensor.dynamic_range)
+    # TODO: handle sum & bias
```

Comment:
is it possible to set bias to zeros to get the matmul dynamic range?

Reply:
good idea for this case. this feature should be supported in the algorithm side. added a comment to make it clear.